worker/worker.go

```diff
@@ -606,6 +609,116 @@ func (w *worker) rhpBroadcastHandler(jc jape.Context) {
 	}
 }
 
+func (w *worker) rhpPruneContractHandlerPOST(jc jape.Context) {
+	// decode fcid
+	var id types.FileContractID
+	if jc.DecodeParam("id", &id) != nil {
+		return
+	}
+
+	// decode potential timeout
+	var pcr api.RHPPruneContractRequest
+	if jc.Decode(&pcr) != nil {
+		return
+	}
+
+	ctx := jc.Request.Context()
+	if pcr.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(jc.Request.Context(), pcr.Timeout)
+		defer cancel()
+	}
+
+	// check if there's prunable data for the contract
+	size, err := w.bus.ContractSize(ctx, id)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch prunable data for contract", err) != nil {
+		return
+	} else if size.Prunable == 0 {
+		return
+	}
+
+	// attach gouging checker
+	gp, err := w.bus.GougingParams(ctx)
+	if jc.Check("could not get gouging parameters", err) != nil {
+		return
+	}
+	ctx = WithGougingChecker(ctx, w.bus, gp)
+
+	// fetch the contract from the bus
+	c, err := w.bus.Contract(ctx, id)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch contract", err) != nil {
+		return
+	}
+
+	// fetch the contract from the bus
+	h, err := w.bus.Host(ctx, c.HostKey)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch host", err) != nil {
+		return
+	}
+
+	// convenience variables
+	hk := c.HostKey
+	hostIP := c.HostIP
+	hostAddr := c.SiamuxAddr
+	hostVersion := h.Settings.Version
+
+	// prune sectors from contracts
+	var deleted int64
+	var toDelete []uint64
+	err = w.withRevision(ctx, defaultRevisionFetchTimeout, id, hk, hostAddr, lockingPriorityPruning, gp.ConsensusState.BlockHeight, func(rev types.FileContractRevision) error {
+		// fetch the roots from the host
+		renterKey := w.deriveRenterKey(c.HostKey)
+		got, err := w.FetchContractRoots(ctx, hostIP, hk, renterKey, id, rev.RevisionNumber, time.Minute)
+		if err != nil {
+			return err
+		}
+
+		// fetch the roots from the bus
+		want, pending, err := w.bus.ContractRoots(ctx, id)
+		if err != nil {
+			return err
+		}
+		keep := make(map[types.Hash256]struct{})
+		for _, root := range append(want, pending...) {
+			keep[root] = struct{}{}
+		}
+
+		// collect indices for roots we want to prune
+		for i, root := range got {
+			if _, wanted := keep[root]; wanted {
+				delete(keep, root) // prevent duplicates
+				continue
+			}
+			toDelete = append(toDelete, uint64(i))
+		}
+		if len(toDelete) == 0 {
+			return fmt.Errorf("no sectors to prune, database holds %d (%d pending), contract contains %d", len(want)+len(pending), len(pending), len(got))
+		}
+
+		// delete the roots from the contract
+		n, err := w.DeleteContractRoots(ctx, hostIP, hostVersion, hk, renterKey, id, rev.RevisionNumber+1, time.Minute, toDelete)
```

Comment:
`DeleteContractRoots` and `FetchContractRoots` need to run within the same rhpv2 session. Meaning we call RPCLock, fetch the roots from the host, fetch the roots from the bus, compute the diff and then delete the roots from the host. All while the contract is locked on the host.

Only then we can be sure that the contract can't accidentally be used in the meantime and no more roots apart from the ones in the database and the in-memory upload cache are going to be added to the contract.

`w.withRevision` shouldn't even be necessary since we already get the locked revision when we perform the rhpv2 handshake.